### ISSUE TITLE
set gradient manually for parameters

### ISSUE THREAD

**Issue by Frozenmad** [PaddlePaddle/Paddle#32500]
- 版本、环境信息：
    1）PaddlePaddle版本：2.0
    4）系统环境：Linux

请问怎么手动设置parameters的gradient，我想完成以下功能：
```python
import paddle
import numpy as np

a = paddle.create_parameter(shape=[4], dtype='float32')
opt = paddle.optimizer.Adam(0.1, parameters=[a])
a.grad = np.array([1.,2.,3.,4.])
opt.step()
```

目前会报错说不能设置attribute

**Comment by hong19860320** [comment]
你指的不是设置梯度，而是如何设置参数'a'的初始值吧？

**Comment by Frozenmad** [comment]
您好，谢谢回复！我是要手动设定a的梯度，并非初始值，提供的例子只是快速展示我要完成的功能的example（实际上我是要在训练模型中在回传的梯度上做点修改，但没找到方法）。

**Comment by hong19860320** [comment]
目前没有提供手动设置梯度的方法哦，为什么要手动设置梯度啊，背景能说下吗？

**Comment by Frozenmad** [comment]
目前在尝试SGLD相关实践：[Bayesian learning via stochastic gradient langevin dynamics](https://www.ics.uci.edu/~welling/publications/papers/stoclangevin_v6.pdf)，如果能改梯度的话好像会方便点。或者有没有办法写自己的optimizer？

**Comment by Frozenmad** [comment]
对了，还有一个场景是accumulative的更新梯度，目前我得解决办法是对loss除以accumulate step。但最好的实现方式感觉是积累后对梯度除以一个accumulate step。

**Comment by www516717402** [comment]
1. paddle修改梯度不清楚，torch里面轻松修改。
2. loss除以accumulate step，这个可以使用loss = loss * accumulate step替代每次backward

**Comment by Frozenmad** [comment]
> 目前没有提供手动设置梯度的方法哦，为什么要手动设置梯度啊，背景能说下吗？

请问后续paddle会支持手动设置梯度吗？

**Comment by hong19860320** [comment]
> > 目前没有提供手动设置梯度的方法哦，为什么要手动设置梯度啊，背景能说下吗？
> 
> 请问后续paddle会支持手动设置梯度吗？

我们记下这个需求，后续统一的排期开发哈~

**Comment by Frozenmad** [comment]
感谢，感觉这个场景需求越来越多了，continual learning里面的ogd方法也需要对gradient进行投影变换之后再更新。

p.s. 给后续有这个需求的开发者：可以使用下面的方法来暂时重设梯度（但可能效率不高）：
```python
parameters = initialize_parameters(...)
grads = calculate_gradients(parameters)
new_grads = transform_gradients(grads)
# Clear grad and backward again using hack loss
for grad, param in zip(new_grads, parameters):
        # clear current grad
        param.clear_grad() 
        # re-compute grad using hack loss
        (param * paddle.to_tensor(grad)).sum().backward()
```

**Comment by Chenfei-Kang** [comment]
啊，找了一个晚上终于有答案了。修改梯度的情形还是有的，比如NLP的对抗训练，需要先储存原有的梯度，把噪声的梯度反向传播一下，再把原有的梯度还原回去，要是改不了的话就不能写了。

**Comment by ZeqiangWangAI** [comment]
> 啊，找了一个晚上终于有答案了。修改梯度的情形还是有的，比如NLP的对抗训练，需要先储存原有的梯度，把噪声的梯度反向传播一下，再把原有的梯度还原回去，要是改不了的话就不能写了。

您好，关于这个对抗训练的部分您有相关代码吗，我最近正在想复现对抗训练，感谢！

**Comment by Chenfei-Kang** [comment]
> > 啊，找了一个晚上终于有答案了。修改梯度的情形还是有的，比如NLP的对抗训练，需要先储存原有的梯度，把噪声的梯度反向传播一下，再把原有的梯度还原回去，要是改不了的话就不能写了。
> 
> 您好，关于这个对抗训练的部分您有相关代码吗，我最近正在想复现对抗训练，感谢！

这个网上一大堆，你搜搜FGM的pytorch咋写的，这个是最简单的，其他的复杂的可能paddle没有对应的接口，我记得当时就只用了这个。
